Add `liftSimplifier` and perform a little restructuring of the codebase

breakerzirconia opened the pull request:
### Scope:

1. Just as in the case of `MonadSMT`, we add `liftSimplifier :: MonadSimplify m => Simplifier a -> m a` to be able to work with nested transformer structures, the base structure of which is `Simplifier a`. With the current infrastructure of the codebase, this introduces a circular import (`Kore.Simplify.Simplify <-> Kore.Simplify.Data`), which can be fixed by moving the definition of `Env`, `Simplify` and `runSimplifier` to `Kore.Simplify.Simplify`. Moreover, we add extra fields to the `Env` datatype and use them in our `MonadSimplify Simplifier` instance to ensure the new circular imports not to occur (otherwise, **tens** of modules are going to be in the state of importing each other).
2. The aforementioned changes break the tests (one of the most common errors include the ambiguity between `Kore.Simplify.Simplify.runSimplifier` and `Test.Kore.Simplify.runSimplifier`), so we fix them.
3. Finally, some of the methods in the `MonadSimplify` typeclass can be defined in terms of `liftSimplifier`, so we push them outside. Additionally, we get rid of the `TestSimplifier` datatype and operate on the original `src` version (`Simplifier`). This reduces a portion of the code in the test suite.

### Estimate:

---

###### Review checklist

The author performs the actions on the checklist. The reviewer evaluates the work and checks the boxes as they are completed.

-   [ ] **Summary.** Write a summary of the changes. Explain what you did to fix the issue, and why you did it. Present the changes in a logical order. Instead of writing a summary in the pull request, you may push a clean Git history.
-   [ ] **Documentation.** Write documentation for new functions. Update documentation for functions that changed, or complete documentation where it is missing.
-   [ ] **Tests.** Write unit tests for every change. Write the unit tests that were missing before the changes. Include any examples from the reported issue as integration tests.
-   [ ] **Clean up.** The changes are already clean. Clean up anything near the changes that you noticed while working. This does not mean only spatially near the changes, but logically near: any code that interacts with the changes!

jberthold commented on kore/test/Test/Kore/Builtin/Builtin.hs:
```diff
@@ -92,7 +92,7 @@ import Kore.Simplify.InjSimplifier
 import Kore.Simplify.Not qualified as Not
 import Kore.Simplify.OverloadSimplifier
 import Kore.Simplify.Pattern qualified as Pattern
-import Kore.Simplify.Simplify hiding (simplifyPattern)
```
(nit) Not really happy with this. Having a different `runSimplifier` in scope is very confusing (it was already confusing before), maybe we should rename the one in `Test.Kore.Simplifier` while we are at it.

jberthold commented on kore/src/Kore/Simplify/Data.hs:
```diff
@@ -131,48 +108,6 @@ traceProfSimplify (Pattern.toTermLike -> termLike) =
-
 -- | Run a simplification, returning the results along all branches.
 runSimplifierBranch ::
     Env ->
```
IMHO this definition should always be adjacent to `runSimplifier` for readability (the two haddock comments connects the two).

jberthold commented on kore/src/Kore/Simplify/Data.hs:
```diff
@@ -20,17 +20,20 @@ module Kore.Simplify.Data (
 
     -- * Re-exports
     MonadSimplify (..),
```
From the list above `-- * Re-exports`, only `evalSimplifier*` still lives in this module. Also, there isn't any "Data" left in `Simplify.Data` (maybe a misnomer in the first place).
Suggest we
1) either move all content from `Kore.Simplify.Data` to `Kore.Simplify`,
2) or make this a new "API" module that contains the runner and utility functions
Blast radius of 2) is of course bigger but once we move things around there is no reason to not do a proper job.

breakerzirconia replied:
By making this a new "API" module you mean renaming the 'Data' module `Kore.Simplify.API` and fixing the respective modules that import this one?

ana-pantilie replied:
I agree with @jberthold, we should probably do what he describes in option 1.

breakerzirconia replied:
@jberthold mentioned moving to `Kore.Simplify`, but there is no such module. Does that mean we need to create this module or was it a typo (was the actual module name `Kore.Simplify.Simplify`)?

jberthold replied:
Sorry, oversight on my end... of course I mean `Kore.Simplify.Simplify`

breakerzirconia replied:
Oh, unfortunately, it seems like numerous circular imports emerge when we try to move everything from `Data` to `Simplify`.
In `Kore.Simplify.Data` we import `Kore.Simplify.TermLike`, and then if we import it in `Kore.Simplify.Simplify`, those tens of circular imports, written in the _Scope_ section of the pull request, make their appearance again.

Should we consider @jberthold's second option instead?

ana-pantilie replied:
Hmm, it's something that will have to be done in order to keep things clean, but I'm not sure how much time it will take. Maybe give it a try and let us know if it ends up being too out of scope for this issue?

jberthold commented on kore/src/Kore/Simplify/Simplify.hs:
```diff
@@ -111,43 +124,69 @@ import Logic
         (MonadTrans t, MonadSimplify n, m ~ t n) =>
-        SideCondition RewritingVariableName ->
-        TermLike RewritingVariableName ->
-        m (OrPattern RewritingVariableName)
-    simplifyTerm sideCondition termLike =
-        lift (simplifyTerm sideCondition termLike)
+        Simplifier a ->
+        m a
+    liftSimplifier = lift . liftSimplifier
+    {-# INLINE liftSimplifier #-}
```
:+1:  DRY!!

jberthold reviewed (commented):
:+1: Makes good sense to refactor this! :sparkles: 
I have a few comments about how to arrange the module contents, though.

jberthold commented on kore/test/Test/Kore/Builtin/Builtin.hs:
```diff
@@ -92,7 +92,7 @@ import Kore.Simplify.InjSimplifier
 import Kore.Simplify.OverloadSimplifier
 import Kore.Simplify.Pattern qualified as Pattern
-import Kore.Simplify.Simplify hiding (simplifyPattern)
+import Kore.Simplify.Simplify hiding (runSimplifier, simplifyPattern)
```
nit) Not really happy with this. Having a different `runSimplifier` in scope is very confusing (it was already confusing before), maybe we should rename the one in `Test.Kore.Simplifier` while we are at it.

breakerzirconia commented:
Right, the following [changes](https://github.com/runtimeverification/haskell-backend/pull/3171/commits/6f7b29f3f40a21430ed8206415b8eff49106ec9f) had been put into effect:
1. Renamed 
  a. `Test.Kore.Simplify.runSimplifier` to `Test.Kore.Simplify.testRunSimplifier`
  b. `Test.Kore.Simplify.runSimplifierBranch` to `Test.Kore.Simplify.testRunSimplifierBranch`
  c. `Test.Kore.Simplify.Condition.runSimplifier` to `Test.Kore.Simplify.Condition.conditionRunSimplifier`
2. Renamed `Kore.Simplify.Data` to `Kore.Simplify.API` everywhere in the project.
3. Gotten rid of `hiding (runSimplifier)` in the test suite.

jberthold reviewed (approved):
LGTM now, nice!
